kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -104,6 +111,14 @@ public Transaction build(Engine engine) {
         boolean isNewTable = snapshot.getVersion(engine) < 0;
         validate(engine, snapshot, isNewTable);
 
+        Metadata  metadata = snapshot.getMetadata().clone();
+        if (tableProperties.isPresent()) {
+            TableConfig.validateProperties(tableProperties.get());
+            Map<String, String> newConfiguration = new HashMap<>(metadata.getConfiguration());
+            newConfiguration.putAll(tableProperties.get());
+            metadata.setConfiguration(newConfiguration);
```

Comment:
Add new method `Metadata.withNewConfiguration(Map<String, String> config)` that returns a copy of the `MEtadata` with new configuration. It also avoids the need for `MEtadata.clone`